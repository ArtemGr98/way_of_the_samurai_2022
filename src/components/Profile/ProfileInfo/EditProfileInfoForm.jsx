import {ErrorMessage, Field, Form, Formik} from "formik";
import * as Yup from "yup";
import {InputForm, ValidationError} from "../../../interface/Form/Form";
import {Button} from "../../../interface/Button/Button";
import {editProfileInfo} from "../../../redux/profile/profile";
import {useDispatch, useSelector} from "react-redux";
import {useEditProfileInfoMutation} from "../../../redux/profile/profileQueryApi";

export const EditProfileInfoForm = ({profileInfo, setEditMode}) => {

    const dispatch = useDispatch()
    const myId = useSelector(state => state.authMe.authMeData.id)

    const [editProfileInfoMutation] = useEditProfileInfoMutation()

    return (
        <Formik
            initialValues={{
                fullName: profileInfo.fullName,
                aboutMe: profileInfo.aboutMe,
                lookingForAJobDescription: profileInfo.lookingForAJobDescription
            }}
            validationSchema={Yup.object().shape({
                fullName: Yup.string()
                    .required(""),
            })}
            onSubmit={async (values, actions) => {
                // const responseData = await dispatch(editProfileInfo(values, myId))
                const responseData = await editProfileInfoMutation(values, myId)
                if (responseData.data.resultCode === 0) {
                    setEditMode(false)
                }
                else {
                    actions.setStatus(responseData.data.messages[0])
                }
                actions.setSubmitting(false);
            }}
        >
            {props => (
                <Form>
                    <div>
                        <div>
                            fullName
                        </div>
                        <Field type="text" name="fullName" component={InputForm}/>
                        <ErrorMessage name="fullName" component={ValidationError}/>
                    </div>
                    <div>
                        <div>
                            aboutMe
                        </div>
                        <Field type="text" name="aboutMe" component={InputForm}/>
                        <ErrorMessage name="aboutMe" component={ValidationError}/>
                    </div>
                    <div>
                        <div>
                            lookingForAJobDescription
                        </div>
                        <Field type="text" name="lookingForAJobDescription" component={InputForm}/>
                        <ErrorMessage name="lookingForAJobDescription" component={ValidationError}/>
                    </div>
                    <ValidationError>
                        {props.status}
                    </ValidationError>
                    <Button type="submit" disabled={!(props.isValid && props.dirty)}>
                        save
                    </Button>
                </Form>
            )}
        </Formik>
    )
}