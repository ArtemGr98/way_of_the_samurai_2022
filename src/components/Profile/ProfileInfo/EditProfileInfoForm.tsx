import {ErrorMessage, Field, Form, Formik} from "formik";
import * as Yup from "yup";
import {InputForm, ValidationError} from "../../../interface/Form/Form";
import {Button} from "../../../interface/Button/Button";
import {useEditProfileInfoMutation} from "../../../redux/profile/profileQueryApi";
import { useAppSelector } from "../../../hook";
import { IProfileInfo } from "../../../redux/profile/profileQueryApiTypes";
import { FC } from "react";

type EditProfileInfoFormProps = {
    profileInfo: IProfileInfo
    setEditMode: (value: boolean) => void
}

export const EditProfileInfoForm: FC<EditProfileInfoFormProps> = ({profileInfo, setEditMode}) => {

    const myId = useAppSelector(state => state.authMe.authMeData.id)

    const [editProfileInfoMutation] = useEditProfileInfoMutation()

    return (
        <Formik
            initialValues={{
                fullName: profileInfo.fullName,
                aboutMe: profileInfo.aboutMe,
                lookingForAJobDescription: profileInfo.lookingForAJobDescription
            }}
            validationSchema={Yup.object().shape({
                fullName: Yup.string()
                    .required(""),
            })}
            onSubmit={(values, actions) => {
                editProfileInfoMutation(values).unwrap()
                .then((payload) => {
                    setEditMode(false)
                    if (payload.resultCode === 0) {
                        setEditMode(false)
                    }
                    else {
                        actions.setStatus(payload.messages[0])
                    }
                })
                .catch((error) => console.log(error))
                actions.setSubmitting(false);
            }}
        >
            {props => (
                <Form>
                    <div>
                        <div>
                            fullName
                        </div>
                        <Field type="text" name="fullName" component={InputForm}/>
                        <ErrorMessage name="fullName" component={ValidationError}/>
                    </div>
                    <div>
                        <div>
                            aboutMe
                        </div>
                        <Field type="text" name="aboutMe" component={InputForm}/>
                        <ErrorMessage name="aboutMe" component={ValidationError}/>
                    </div>
                    <div>
                        <div>
                            lookingForAJobDescription
                        </div>
                        <Field type="text" name="lookingForAJobDescription" component={InputForm}/>
                        <ErrorMessage name="lookingForAJobDescription" component={ValidationError}/>
                    </div>
                    <ValidationError>
                        {props.status}
                    </ValidationError>
                    <Button type="submit" disabled={!(props.isValid && props.dirty)}>
                        save
                    </Button>
                </Form>
            )}
        </Formik>
    )
}