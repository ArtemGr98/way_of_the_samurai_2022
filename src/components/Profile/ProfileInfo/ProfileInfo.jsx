import background from "../../../img/Profile/background.jpg";
import profileImg from "../../../img/Profile/profileImg.png";
import styled from "styled-components";
import Loader from "../../common/Loader/Loader";
import React from "react";

const ProfileImg = styled.div`
  width: 100%;
  height: 300px;

  img {
    width: 100%;
    height: 100%;
  }
`
const ProfileInfoWrapper = styled.div`
  margin-top: 20px;
  display: flex;
  max-height: 300px;
`
const ProfileAva = styled.div`
  width: 20%;
  height: 100%;

  img {
    width: 100%;
    height: 100%;
  }
`
const ProfileDescription = styled.div`
  width: 80%;
  padding-left: 20px
`


class ProfileInfo extends React.Component {
    state = {
        editMode: false,
        status: this.props.status
    }

    editToggle = () => {
        this.setState({
            editMode: !this.state.editMode
        }, () => {
            if (!this.state.editMode) {
                this.props.updateStatus(this.state.status)
            }
        })
    }

    cancel = () => {
        this.setState({
            editMode: false
        })
    }

    onChangeStatus = (e) => {
        this.setState({
            status: e.currentTarget.value
        })
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (prevProps.status !== this.props.status) {
            this.setState({
                status: this.props.status
            })
        }
    }

    render() {
        if (!this.props.profile) {
            return <Loader/>
        }
        return (
            <div>
                <ProfileImg>
                    <img src={background} alt="background"/>
                </ProfileImg>

                <ProfileInfoWrapper>
                    <ProfileAva>
                        <img
                            src={this.props.profile.photos.large ? this.props.profile.photos.large : profileImg}
                            alt="profileImg"/>
                    </ProfileAva>
                    <ProfileDescription>
                        <div>
                            {this.props.profile.fullName}
                        </div>
                        <div>
                            {this.state.editMode ?
                                <div>
                                    <input autoFocus={true} type="text"
                                           value={this.state.status} onChange={this.onChangeStatus}/>
                                    <button onClick={this.cancel}>
                                        cancel
                                    </button>
                                </div>
                                :
                                <span>
                                    {this.props.status || "no status"}
                                </span>
                            }
                            <button onClick={this.editToggle}>
                                edit
                            </button>
                        </div>
                        <div>
                            {this.props.profile.aboutMe ? this.props.profile.aboutMe : "aboutMe"}
                        </div>

                    </ProfileDescription>
                </ProfileInfoWrapper>
            </div>
        )
    }
}

export default ProfileInfo