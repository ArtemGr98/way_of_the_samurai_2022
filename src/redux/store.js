import {ACTIVE_CHAT_ID, ADD_MESSAGE, ADD_POST, UPDATE_MESSAGE, UPDATE_POST} from "../actions/actionType";

export const store = {
    _subscriber: () => {
        console.log("no subscribers")
    },
    _state: {
        messages: {
            chatListData: [
                {name: "name1", id: 1},
                {name: "name2", id: 2},
                {name: "name3", id: 3},
                {name: "name4", id: 4},
            ],
            chatData: [
                {
                    chatId: 1,
                    chatText: [
                        {text: "text1", id: 1, position: "start"},
                        {text: "text1", id: 2, position: "end"},
                        {text: "text1", id: 3, position: "end"},
                        {text: "text1", id: 4, position: "start"},
                    ]
                },
                {
                    chatId: 2,
                    chatText: [
                        {text: "text2", id: 1, position: "start"},
                        {text: "text2", id: 2, position: "end"},
                        {text: "text2", id: 3, position: "end"},
                        {text: "text2", id: 4, position: "start"},
                    ]
                },
            ],
            activeChat: '',
            textareaState: '',
        },
        profile: {
            postData: [
                {
                    text: "post text",
                    like: 4,
                    dislike: 1,
                    id: 1
                },
                {
                    text: "post text2",
                    like: 43,
                    dislike: 12,
                    id: 2
                },
                {
                    text: "post text3",
                    like: 14,
                    dislike: 21,
                    id: 3
                },
            ],
            textareaState: '',
        },
    },

    getState() {
        return this._state
    },
    subscribe(observer) {
        this._subscriber = observer
    },

    dispatch(action) {
        if (action.type === UPDATE_POST) {
            this._state.profile.textareaState = action.postText
            this._subscriber(this)
        }
        else if (action.type === ADD_POST) {
            const state = this.getState()
            const statePost = state.profile.postData

            const post = {
                text: state.profile.textareaState,
                like: 0,
                dislike: 0,
                id: statePost.length + 1
            }
            statePost.push(post)
            state.profile.textareaState = ''
            this._subscriber(this)
        }
        else if (action.type === UPDATE_MESSAGE) {
            const state = this.getState()
            state.messages.textareaState = action.messageText
            this._subscriber(this)
        }
        else if (action.type === ADD_MESSAGE) {
            const state = this.getState()
            const chatArr = state.messages.chatData
            const id = state.messages.activeChat
            const currentChat = chatArr.filter((chat) => chat.chatId === id)
            console.log(currentChat)
            const chatMessage = {
                text: state.messages.textareaState,
                id: 1,
                position: "end"
            }
            currentChat[0].chatText.push(chatMessage)
            state.messages.textareaState = ''
            this._subscriber(this)
        }
        else if (action.type === ACTIVE_CHAT_ID) {
            this.getState().messages.activeChat = action.id
        }
    },
}


