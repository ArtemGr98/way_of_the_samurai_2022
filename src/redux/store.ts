import { combineReducers, configureStore } from "@reduxjs/toolkit";
import authMeReducer from "./authMe/authMe";
import initAppReducer from "./initApp/initApp";
import profileReducer from "./profile/profile";
import usersReducer from "./users/users";
import messagesReducer from "./messages/messages";
import {profileQueryApi} from "./profile/profileQueryApi";

const rootReducer = combineReducers({
    profile: profileReducer,
    initApp: initAppReducer,
    authMe: authMeReducer,
    users: usersReducer,
    messages: messagesReducer,
    [profileQueryApi.reducerPath]: profileQueryApi.reducer,
})

export const store = configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware().concat(profileQueryApi.middleware)
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch