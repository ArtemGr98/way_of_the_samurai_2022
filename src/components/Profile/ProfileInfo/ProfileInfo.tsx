import styled from "styled-components";
import Loader from "../../common/Loader/Loader";
import {ChangeEvent, FC, useState} from "react";
import {useEffect} from "react";
import {EditProfileInfoForm} from "./EditProfileInfoForm";
import {
    useGetProfileInfoQuery,
    useGetStatusQuery, useSavePhotoMutation,
    useUpdateStatusMutation
} from "../../../redux/profile/profileQueryApi";


const ProfileInfoWrapper = styled.div`
  margin-top: 20px;
  display: flex;
  max-height: 300px;
`
const ProfileAva = styled.div`
  width: 20%;
  height: 100%;

  img {
    width: 100%;
    height: 100%;
  }
`
const ProfileDescription = styled.div`
  width: 80%;
  padding-left: 20px
`
const ProfileTop = styled.div`
  position: relative;
`
const EditBtn = styled.button`
  position: absolute;
  top: 0;
  right: 0;
`

interface ProfileInfoProps {
    userId: string;
    isMyProfile: boolean;
}

const ProfileInfo: FC<ProfileInfoProps> = ({userId, isMyProfile}) => {

    const {data: profileInfo, isLoading, isError, error} = useGetProfileInfoQuery(userId)
    const {data: status} = useGetStatusQuery(userId)

    const [updateStatusMutation] = useUpdateStatusMutation()
    const [savePhoto] = useSavePhotoMutation()

    const [editMode, setEditMode] = useState(false)
    const [editModeStatus, setEditModeStatus] = useState(false)
    const [stateStatus, setStatus] = useState(status)

    useEffect(() => {
        setStatus(status)
    }, [status])

    const editToggleStatus = () => {
        setEditModeStatus(!editModeStatus)
        if (editModeStatus && stateStatus) {
            updateStatusMutation(stateStatus).unwrap()
        }
    }

    const onChangePhoto = async (e: ChangeEvent<HTMLInputElement>) => {
        let target = e.target
        const files = target.files

        if (files) {
            await savePhoto(files[0])
            target.value = ''
        }
    }

    if (isLoading) {
        return <Loader/>
    }

    if (isError) {
        return <div>Error</div>
    }
    
    return (
        <ProfileTop>
            {editMode && profileInfo ? <EditProfileInfoForm profileInfo={profileInfo} setEditMode={setEditMode}/> :
                <ProfileInfoWrapper>
                    <ProfileAva>
                        <img
                            src={profileInfo?.photos?.large && profileInfo.photos.large}
                            alt="profileImg"/>
                        {isMyProfile && <input type="file" onChange={onChangePhoto}/>}
                    </ProfileAva>
                    <ProfileDescription>
                        <div>
                            {profileInfo?.fullName}
                        </div>
                        <div>
                            {editModeStatus ?
                                <div>
                                    <input autoFocus={true} type="text"
                                           value={stateStatus}
                                           onChange={(e) => setStatus(e.currentTarget.value)}/>
                                    <button onClick={() => setEditModeStatus(false)}>
                                        cancel
                                    </button>
                                </div>
                                :
                                <span>
                                {status || "no status"}
                            </span>}
                            {isMyProfile && <button onClick={editToggleStatus}>
                                edit status
                            </button>}
                        </div>
                        <div>
                            {profileInfo?.aboutMe || "aboutMe"}
                        </div>
                    </ProfileDescription>
                </ProfileInfoWrapper>
            }

            {isMyProfile && <EditBtn onClick={() => setEditMode(!editMode)}>edit</EditBtn>}
        </ProfileTop>
    )
}

export default ProfileInfo