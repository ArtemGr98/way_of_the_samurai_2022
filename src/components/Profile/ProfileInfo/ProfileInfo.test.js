import TestRenderer from 'react-test-renderer';
import ProfileInfo from "./ProfileInfo";
import {Provider} from "react-redux";
import store from "../../../redux/store";

const testRenderer = TestRenderer.create(<Provider store={store}><ProfileInfo /></Provider> );
const testInstance = testRenderer.root;

test('renders correctly', () => {
    expect(testRenderer).toMatchSnapshot()
})