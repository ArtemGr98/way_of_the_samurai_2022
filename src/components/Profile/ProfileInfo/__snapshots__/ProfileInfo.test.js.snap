// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-bczRLJ fCXZKv"
>
  <img
    alt="loader"
    src="loader2.gif"
  />
</div>
`;
